src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,227 @@
+  Log::Fatal << "Increase population size." << std::endl;
+
+  // Get the number of functions to iterate.
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1].
+  population = arma::randu(iterate.n_rows, populationSize);
+
+  // initializing helper variables.
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness = 0;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+      << std::endl;
+
+  // Find the fitness before optimization using given iterate parameters.
+  for (size_t j = 0; j < numFun; j++)
+    lastBestFitness += function.Evaluate(iterate, j);
+
+  // Iterate until maximum number of generations is obtained.
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // Calculating fitness values of all candidates.
+    for (size_t i = 0; i < populationSize; i++)
+    {
+       // Select a candidate and insert the parameters in the function.
+       iterate = population.col(i);
+
+       // Find fitness of candidate.
+       for (size_t j = 0; j < numFun; j++)
+         fitness += function.Evaluate(iterate, j);
+
+       // Save fitness value of the evaluated candidate.
+       fitnessValues[i] = fitness;
+       fitness = 0;
+    }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+          << fitnessValues.min() << std::endl;
+
+      // Create next generation of species.
+      Reproduce();
+
+      // Check for termination criteria.
+      if (finalValue != DBL_MIN && finalValue >= fitnessValues.min())
+      {
+          Log::Info << "Terminating. Given fitness criteria " << finalValue
+              << " > " << fitnessValues.min() << std::endl;
+          break;
+      }
+
+      // Check for termination criteria.
+      if (objectiveChange != DBL_MIN &&
+        (lastBestFitness - fitnessValues.min()) < objectiveChange)
+      {
+          Log::Info << "Terminating. Fitness History change "
+              << (lastBestFitness - fitnessValues.min())
+          << " < " << objectiveChange << std::endl;
+          break;
+      }
+
+      // Store the best fitness of present generation.
+      lastBestFitness = fitnessValues.min();
+  }
+
+  // Set the best candidate into the network parameters.
+  iterate = population.col(index(0));
+
+  // Find the objective function value from the best candidate found.
+  for (size_t j = 0; j < numFun; j++)
+      fitness += function.Evaluate(iterate, j);
+
+  return fitness;
+}
+
+//! Reproduce candidates to create the next generation.
+void CNE::Reproduce()
+{
+  // Sort fitness values. Smaller fitness value means better performance.
+  index = arma::sort_index(fitnessValues);
+
+  // First parent.
+  size_t mom;
+
+  // Second parent.
+  size_t dad;
+
+  for (size_t i = numElite; i < populationSize - 1; i++)
+  {
+    // Select 2 different parents from elite group randomly [0, numElite).
+    do {
+         mom = mlpack::math::RandInt(0, numElite);
```

Comment:
to use next or previous index will then call for check if index is out of bounds [0, numElite)